src/coreclr/vm/reflectioninvocation.cpp

```diff
@@ -164,41 +144,24 @@ FCIMPL3(Object*, ReflectionInvocation::AllocateValueType, ReflectClassBaseObject
     gc.obj = gc.value;
     gc.refTargetType = (REFLECTCLASSBASEREF)ObjectToOBJECTREF(pTargetTypeUNSAFE);
 
+    HELPER_METHOD_FRAME_BEGIN_RET_PROTECT(gc);
+
     TypeHandle targetType = gc.refTargetType->GetType();
 
-    HELPER_METHOD_FRAME_BEGIN_RET_PROTECT(gc);
-    CorElementType targetElementType = targetType.GetSignatureCorElementType();
-    if (InvokeUtil::IsPrimitiveType(targetElementType) || targetElementType == ELEMENT_TYPE_VALUETYPE)
-    {
-        MethodTable* allocMT = targetType.AsMethodTable();
+    // This method is only intended for value types; it is not called directly by any public APIs
+    // so we don't expect validation issues here.
+    _ASSERTE(InvokeUtil::IsPrimitiveType(targetType.GetSignatureCorElementType()) ||
+        targetType.GetSignatureCorElementType() == ELEMENT_TYPE_VALUETYPE);
 
-        if (allocMT->IsByRefLike()) {
-            COMPlusThrow(kNotSupportedException, W("NotSupported_ByRefLike"));
-        }
+    MethodTable* allocMT = targetType.AsMethodTable();
+    _ASSERTE(!allocMT->IsByRefLike());
 
-        if (gc.value != NULL)
-        {
-            // ignore the type of the incoming box if fForceTypeChange is set
-            // and the target type is not nullable
-            if (!fForceTypeChange || Nullable::IsNullableType(targetType))
-                allocMT = gc.value->GetMethodTable();
-        }
+    gc.obj = allocMT->Allocate();
+    _ASSERTE(gc.obj != NULL);
 
-        // for null Nullable<T> we don't want a default value being created.
-        // just allow the null value to be passed, as it will be converted to
-        // a true nullable
-        if (!(gc.value == NULL && Nullable::IsNullableType(targetType)))
-        {
-            // boxed value type are 'read-only' in the sence that you can't
-            // only the implementor of the value type can expose mutators.
-            // To insure byrefs don't mutate value classes in place, we make
-            // a copy (and if we were not given one, we create a null value type
-            // instance.
-            gc.obj = allocMT->Allocate();
-
-            if (gc.value != NULL)
-                    CopyValueClass(gc.obj->UnBox(), gc.value->UnBox(), allocMT);
-        }
+    if (gc.value != NULL) {
```

Comment:
This cannot be ever null (the allocator throws OOM instead of returning null).

Reply:
`gc.value` here is the incoming value that can be `null`, and not the result of the allocation which is `gc.obj`. Basically this method has two modes - one just to allocate a value type where `gc.value` is null, and another to allocate plus copy an existing boxed value that is in `gc.value`.